MetadataExtractor/Formats/Gif/GifReader.cs

```diff
@@ -332,17 +333,21 @@ private static GifImageDirectory ReadImageBlock(SequentialReader reader)
         private static byte[] GatherBytes(SequentialReader reader)
         {
             var bytes = new MemoryStream();
-            var buffer = new byte[257];
+            var buffer = ArrayPool<byte>.Shared.Rent(257);
 
             while (true)
             {
-                var b = reader.GetByte();
-                if (b == 0)
-                    return bytes.ToArray();
-                buffer[0] = b;
-                reader.GetBytes(buffer, 1, b);
-                bytes.Write(buffer, 0, b + 1);
+                var len = reader.GetByte();
+                if (len == 0)
+                    break;
+                buffer[0] = len;
+                reader.GetBytes(buffer, offset: 1, count: len);
+                bytes.Write(buffer, 0, len + 1);
             }
+
+            ArrayPool<byte>.Shared.Return(buffer);
```

Comment:
Could we also kill the`var bytes = new MemoryStream();` allocation, and write directly into the buffer (keeping track of the written bytes)?

Reply:
The issue is we don't know how large the stream will be in the end. Each iteration of the loop will add another chunk of bytes, and the result needs to be trimmed to size (with the current API). It's a pretty bizarre encoding format (XMP in GIF &mdash; see https://dl.photoprism.app/pdf/20120101-Adobe_XMP_Specification_Part_3.pdf section 1.1.2 on page 11).

Reply:
I pushed some more improvements in GifReader.